spyql/parser.py

```diff
@@ -91,37 +91,20 @@ def parse_structure(query: str):
     query_struct_keyword_dict: Dict[str, List[str]] = {
         mk[0]: mk for mk in [k.split() for k in query_struct_keywords]
     }
-    present_keyword = None
 
     validator = KeywordOrderValidator(query_struct_keywords)
-
-    i = 0
-    while i < len(tokens):
-        token = tokens[i]
-
-        keyword_parts = query_struct_keyword_dict.get(token.lower())
-        if keyword_parts and keyword_parts == [
-            t.lower() for t in tokens[i : i + len(keyword_parts)]
-        ]:
-            present_keyword = " ".join(keyword_parts)
-            validator.run(present_keyword)
-            query_struct[present_keyword] = ""
-            i += len(keyword_parts)
-
-        # For not-keyword token
-        else:
-            if not present_keyword:
-                log.user_error(
-                    "could not parse query",
-                    SyntaxError(f"misplaced '{token}' clause"),
-                )
-
-            if query_struct[present_keyword]:
-                query_struct[present_keyword] += f" {token}"
-            else:
-                query_struct[present_keyword] += token
-
-            i += 1
+    keyword_positions = [
+        (i, query_struct_keyword_dict[token.lower()])
+        for i, token in enumerate(tokens)
+        if token.lower() in query_struct_keyword_dict
+    ]
+
+    for i, (pos, kw) in enumerate(keyword_positions):
+        validator.run(" ".join(kw))
```

Comment:
we still need to validate if `token[pos:pois+len(kw)] == kw` (I haven't included conversion of tokens to lower case, but it is required). 
`keyword_positions` is in fact a list of candidates that might have false positives because it is based only on the first word of the keyword. This changes the logic a bit, since instead of looking forward for next keyword, it is easier to track the last valid keyword.

Examples of queries that have unexpected behaviour with this implementation:

```sh
$ echo "order, value
123, 100
124, 150
125, 88
126, 201" > test.csv

$ python3 -m spyql.cli "SELECT value, order FROM csv"        
ERROR	could not parse query
SyntaxError: misplaced 'from' clause

$ python3 -m spyql.cli "SELECT col1 FROM range(3,0,-1) ORDER 1"
col1
3
2
1

$ python3 -m spyql.cli "SELECT col1 FROM range(3,0,-1) ORDER BUZZZZZ 1"
col1
1
2
3
```

Reply:
> we still need to validate if token[pos:pois+len(kw)] == kw (I haven't included conversion of tokens to lower case, but it is required).

You are right, but I think that the validation should be in `keyword_positions`. I suggest the following logic (already checked the first query succeeds and the second and third queries above failed by SyntaxError).

```python
  matchable_tokens: List[Optional[List[str]]] = [
      query_struct_keyword_dict.get(token.lower()) for token in tokens
  ]
  keyword_positions = [
      (i, matchable_token)
      for i, matchable_token in enumerate(matchable_tokens)
      if matchable_token
      and [t.lower() for t in tokens[i : i + len(matchable_token)]] == matchable_token
  ]
```

I want your opinion.

Reply:
Looks good to me! 👍😄